java/src/main/native/src/RmmJni.cpp

```diff
@@ -26,12 +26,120 @@ using rmm::mr::device_memory_resource;
 
 namespace {
 
+// Alignment to which the RMM memory resource will round allocation sizes
+constexpr std::size_t RMM_ALLOC_SIZE_ALIGNMENT = 512;
+
 constexpr char const* RMM_EXCEPTION_CLASS = "ai/rapids/cudf/RmmException";
 
+/**
+ * @brief An RMM device memory resource that delegates to another resource
+ * while tracking the amount of memory allocated.
+ *
+ * @tparam Upstream Type of memory resource that will be wrapped.
+ * @tparam size_align The size to which all allocation requests are
+ * aligned. Must be a value >= 1.
+ */
+template <typename Upstream>
+class tracking_resource_adaptor final : public device_memory_resource {
+public:
+  /**
+   * @brief Constructs a new tracking resource adaptor that delegates to
+   * `mr` for all allocation operations while tracking the amount of memory
+   * allocated.
+   *
+   * @param mr The resource to use for memory allocation operations.
+   * @param size_alignment The alignment to which the `mr` resource will
+   * round up all memory allocation size requests.
+   */
+  tracking_resource_adaptor(Upstream* mr, std::size_t size_alignment)
+      : resource{mr}, size_align{size_alignment} {}
+
+  Upstream* get_wrapped_resource() {
+    return resource;
+  }
+
+  std::size_t get_total_allocated() {
+    std::lock_guard<std::mutex> lock(size_map_mutex);
+    return total_allocated;
+  }
+
+private:
+  Upstream* const resource;
+  std::size_t const size_align;
+  std::size_t total_allocated{0};
+
+  // map and associated lock to track memory sizes by address
+  // TODO: This should be removed when rmm::alloc and rmm::free are removed and the size parameter
+  //       for do_deallocate can be trusted. If map and mutex are removed then total_allocated
+  //       should be updated to be atomic.
+  std::unordered_map<void*, std::size_t> size_map{};
+  std::mutex size_map_mutex{};
+
+
+  void* do_allocate(std::size_t num_bytes, cudaStream_t stream) override {
+    // adjust size of allocation based on specified size alignment
+    num_bytes = (num_bytes + size_align - 1) / size_align * size_align;
+
+    std::lock_guard<std::mutex> lock(size_map_mutex);
+
+    auto result = resource->allocate(num_bytes, stream);
+    if (result) {
+      total_allocated += num_bytes;
+      size_map[result] = num_bytes;
+    }
+    return result;
+  }
+
+  void do_deallocate(void* p, std::size_t size, cudaStream_t stream) override {
+    std::lock_guard<std::mutex> lock(size_map_mutex);
+
+    resource->deallocate(p, size, stream);
+
+    if (p) {
```

Comment:
same here, unless we are worried about `deallocate`, why not take the lock right when trying to update the map?

Reply:
The problem is this theoretical scenario:
1. Thread 1 allocates 512 bytes at address X
1. Thread 1 later calls deallocate which releases the memory but we then go to grab lock.
1. Before grabbing the lock, Thread 1 either blocks or loses the CPU
1. Thread 2 comes in and allocates 5120 bytes with the same resulting address X, successfully grabbing the lock and updating the map.  That update smashed the size we had stored in the map for Thread 1's deallocate still being processed.
1. Thread 1 finally regains CPU, successfully grabs the lock, and sees _the wrong size for the allocation being freed_.  It removes the entry in the map, and we've lost the 5120 byte allocation completely in the tracking.

Reply:
Ok yeap that makes total sense. Thank you!